scipy/interpolate/_isotonic.py

```diff
@@ -0,0 +1,123 @@
+from __future__ import annotations
+from typing import TYPE_CHECKING
+
+import numpy as np
+
+from scipy.optimize import isotonic_regression
+
+if TYPE_CHECKING:
+    import numpy.typing as npt
+
+
+__all__ = ["IsotonicInterpolator"]
+
+
+# TODO: Should this inherit from _Interpolator1D?
+class IsotonicInterpolator():
+    """Nonparametric isotonic interpolation via PAVA.
+
+    This class provides nonparametric monotonically increasing interpolation
+    via the pool adjacent violators algorithm (PAVA).
+
+    This interpolator first sorts the values of ``y`` according to the order
```

Comment:
Maybe require x and y to be sorted. Otherwise if it sorting happens to be a perf issue, we wouldn't be able to undo it because backwards compat. Adding functionality is easier than removing.

Reply:
Frankly, I don't think so. an argsort call for a user not that much of a burden. If and only if this happens to be a big deal indeed, we can always add a switch.

Reply:
I see the different perspectives now a bit better.

From a statistician / machine learning perspective, you learn to predict (some statistical functional of y given x. That's why in scikit-learn, we always have `estimator.fit(X, y)` and then `estimator.predict(X)`. The sorting of y on x is, in a sense, a detail of that estimator (which wants to call pava). (Almost?) No other estimator needs to sort. Therefore a user should not be bothered with it.

But let's focus on `isotonic_regression` and `pava` and decide later what happens with this not-an-interpolator class.